File: src/main/java/rbq/wtf/lycoris/client/gui/ClickGUI/Component/Component.java
package rbq.wtf.lycoris.client.gui.ClickGUI.Component;

public class Component {
    private float height = 0;
    private float width = 0;

    public Component() {
    }

    public void render() {
    }

    public void keyTyped(char typedChar, int keyCode) {
    }

    public void updateComponent(float X, float Y, int mouseX, int mouseY) {
    }

    public void mouseClicked(int mouseX, int mouseY, int mouseButton) {
    }

    public float getHeight() {
        return height;
    }

    public void setHeight(float height) {
        this.height = height;
    }

    public float getWidth() {
        return width;
    }

    public void setWidth(float width) {
        this.width = width;
    }

    public boolean isHovered(float x, float y, float x2, float y2, int mouseX, int mouseY) {
        return (float) mouseX >= x && (float) mouseX <= x2 && (float) mouseY >= y && (float) mouseY <= y2;
    }

    public void mouseReleased(int mouseButton) {

    }
}

File: src/main/java/rbq/wtf/lycoris/client/clickgui/component/Component.kt
package rbq.wtf.lycoris.client.clickgui.component

import rbq.wtf.lycoris.client.Client
import rbq.wtf.lycoris.client.clickgui.ClickGUI

abstract class Component(val clickGUI: ClickGUI) : IComponent

abstract class ClickableComponent(val clickGUI: ClickGUI) : IClickableComponent

interface IClickableComponent : IComponent {
    override var hoverOffsetX: Int
    override var hoverOffsetY: Int

    override val hoverStartX: Int
        get() = (x - hoverOffsetX).toInt()
    override val hoverStartY: Int
        get() = (y - hoverOffsetY).toInt()

    override val hoverEndX: Int
        get() = (endX + hoverOffsetX).toInt()
    override val hoverEndY: Int
        get() = (endY + hoverOffsetY).toInt()
}

interface IComponent {
    val startX: Float
        get() = Client.clickGUI.startX
    val startY: Float
        get() = Client.clickGUI.startY
    var offsetX: Float
    var offsetY: Float
    var width: Float
    var height: Float

    val x: Float
        get() = startX + offsetX
    val y: Float
        get() = startY + offsetY

    val endX: Float
        get() = x + width
    val endY: Float
        get() = y + height

    val hoverStartX: Int
        get() = x.toInt()
    val hoverStartY: Int
        get() = y.toInt()

    val hoverEndX: Int
        get() = endX.toInt()
    val hoverEndY: Int
        get() = endY.toInt()

    val hoverOffsetX: Int
        get() = 0
    val hoverOffsetY: Int
        get() = 0

    fun render(mouseX: Int, mouseY: Int, partialTicks: Float)
    fun keyTyped(typedChar: Char, keyCode: Int) {}
    fun updateComponent(mouseX: Int, mouseY: Int) {}
    fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {}
    fun mouseReleased(mouseX: Int, mouseY: Int, mouseButton: Int) {}
    fun mouseWheelScroll(mouseX: Int, mouseY: Int, mouseWheel: Int) {}

//    fun updateStart(startX: Float, startY: Float) {
//        this.startX = startX
//        this.startY = startY
//    }

    fun updateOffset(offsetX: Float, offsetY: Float) {
        this.offsetX = offsetX
        this.offsetY = offsetY
    }

    fun isHovered(startX: Int, startY: Int, endX: Int, endY: Int, mouseX: Int, mouseY: Int): Boolean {
        return mouseX in startX - hoverOffsetX..endX + hoverOffsetX && mouseY in startY - hoverOffsetY..endY + hoverOffsetY
    }

    fun isHovered(mouseX: Int, mouseY: Int): Boolean {
        return mouseX in hoverStartX..hoverEndX && mouseY in hoverStartY..hoverEndY
    }
}